rstudio/rstudio
# Menus now load keyboard shortcuts for App Commands.

See: https://github.com/rstudio/rstudio/issues/5256

### Intent

Note: For the sake of brevity and ease-of-review, this PR addresses menus only. Displaying custom shortcuts in other locations will have to be evaluated on a case-by-case basis. We can have a discussion about whether performing some kind of event trigger-based update to force tooltips to refresh is worth the deep-dive. There are a couple of ways to design that.

An issue (5256) was identified where custom shortcuts set through the "Modify Keyboard Shortcuts" do not appear next to their commands in menus.

The title of the original ticket mentions the command palette as well, but that has been previously addressed and fixed.

### Approach

I've added a custom keybinding shortcut property to the `AppCommand` class which override the display of the default shortcut. Bindings loaded from the config file are considered "custom," which do not overwrite the default shortcut value on the AppCommand itself (but they still change the binding from default). Tooltips still do not immediately update, as they are rendered in when the pane itself is rendered.

Refreshing the page works with varying success, but the browser caches most of the DOM. I could only get source editor tooltips to update by closing and then re-opening a source editor, for example. This will need a more in-depth approach to updating using more event listeners, and that could complicate things.

An earlier approach I took worked well for customizing AppCommands that had no shortcut to begin with, but would override commands with an initial value, such that resetting would no longer work. The current approach makes it possible to reset correctly.

### Automated Tests

An automated browser test would be required to input the keys and then check the visual outcome in a menu.

### QA Notes

**Verify Constraints:**

- The command must appear in a menu
- The command be available to assign a custom shortcut to. Some commands have custom shortcuts disabled.

**Test:**

1. Test changing commands
  - Open the Keyboard Shortcut editor and make changes to a blank and non-blank (with defaults) command.
  - Apply the changes
  - Verify that the items in their menus display the new shortcuts
  - Verify that the app command palette shows the correct new values

2. Test resetting
 - Reset changes
 - Verify that the items have returned to their previous states, blank or otherwise
 - Verify that the app command palette shows the correct values

I've also fixed a minor issue related to the visual display of changing commands in the Shortcuts Editor, so keep an eye out for similar tickets about that: Keyboard shortcuts are now compared correctly, so the editor no longer displays a highlighted row just by clicking in and clicking out again. Similarly, if you input the same command before you made changes, it resets the changed state visually by removing the highlight.

1. Verify current behavior
  - Open the Keyboard Shortcut editor and click in to, then out of any row
  - The row should be highlighted in green as if it has a change, even if there is no change
  - Change any existing keyboard shortcut, overriding the default, and then change it back manually
  - The row should remain highlighted as if has a change, even though the value matches the old value

2. Verify fixed behavior
  - Open the Keyboard Shortcut editor and click in to, then out of any row
  - The row should not be highlighted, correctly showing no change
  - Change any existing keyboard shortcut, overriding the default, and then change it back manually
  - The row should not be highlighted, correctly showing no change from the original value

### Checklist
